programs/lz4io.c

```diff
@@ -1278,82 +1278,118 @@ int LZ4IO_decompressMultipleFilenames(LZ4IO_prefs_t* const prefs,
 } LZ4IO_cFileInfo_t;
 
-#define LZ4IO_INIT_CFILEINFO   { LZ4F_INIT_FRAMEINFO, NULL, 0ULL }
-
+#define LZ4IO_INIT_CFILEINFO   { NULL, 0ULL, 0, LZ4IO_INIT_FRAMEINFO, 1, 1, 1 }
 
 typedef enum { LZ4IO_LZ4F_OK, LZ4IO_format_not_known, LZ4IO_not_a_file } LZ4IO_infoResult;
 
-/* This function is limited,
- * it only works fine for a file consisting of a single valid frame using LZ4 Frame specification.
- * It will not look at content beyond first frame header.
- * It's also unable to parse legacy frames, nor skippable ones.
- *
- * Things to improve :
- * - check the entire file for additional content after first frame
- *   + combine results from multiple frames, give total
- * - Optional :
- *  + report nb of blocks, hence max. possible decompressed size (when not reported in header)
- */
-static LZ4IO_infoResult
-LZ4IO_getCompressedFileInfo(LZ4IO_cFileInfo_t* cfinfo, const char* input_filename)
-{
-    LZ4IO_infoResult result = LZ4IO_format_not_known;  /* default result (error) */
-
-    if (!UTIL_isRegFile(input_filename)) return LZ4IO_not_a_file;
-    cfinfo->fileSize = UTIL_getFileSize(input_filename);
+static const char * LZ4IO_frameTypeNames[]={"LZ4Frame", "LegacyFrame", "SkippableFrame" };
 
-    /* Get filename without path prefix */
-    {   const char* b = strrchr(input_filename, '/');
-        if (!b) {
-            b = strrchr(input_filename, '\\');
-        }
-        if (b && b != input_filename) {
-            b++;
-        } else {
-            b = input_filename;
+/* Read block headers and skip block data
+   Return total blocks size for this frame including headers,
+   block checksums and content checksums.
+   returns 0 in case it can't succesfully skip block data.
+   Assumes SEEK_CUR after frame header.
+ */
+static unsigned long long LZ4IO_skipBlocksData(FILE* finput,
+                                               const LZ4F_blockChecksum_t blockChecksumFlag,
+                                               const LZ4F_contentChecksum_t contentChecksumFlag){
+  unsigned char blockInfo[LZ4F_BLOCK_HEADER_SIZE];
+  unsigned long long totalBlocksSize = 0;
+  for(;;){
+    if (!fread(blockInfo, 1, LZ4F_BLOCK_HEADER_SIZE, finput)){
+      if (feof(finput)) return totalBlocksSize;
+      return 0;
+    }
+    totalBlocksSize += LZ4F_BLOCK_HEADER_SIZE;
+    {
+      const unsigned long nextCBlockSize = LZ4IO_readLE32(&blockInfo) & 0x7FFFFFFFU;
+      const unsigned long nextBlock = nextCBlockSize + (blockChecksumFlag * LZ4F_BLOCK_CHECKSUM_SIZE);
+      if (nextCBlockSize == 0){
+        /* Reached EndMark */
+        if(contentChecksumFlag){
+          /* Skip content checksum */
+          if(fseek(finput, LZ4F_CONTENT_CHECKSUM_SIZE, SEEK_CUR) != 0){
+            return 0;
+          }
+          totalBlocksSize += LZ4F_CONTENT_CHECKSUM_SIZE;
         }
-        cfinfo->fileName = b;
+        break;
+      }
+      totalBlocksSize += nextBlock;
+      /* skip to the next block */
+      if (fseek(finput, nextBlock, SEEK_CUR) != 0){
```

Comment:
`fseek()` : a potential issue here : "normal" `fseek()` in 32-bit mode can have difficulty working properly with large files (> 4 GB). The impact is more or less implementation dependent.

1) Try to re/produce the issue. Typically, use a 32-bit Linux image in a VM. Alternatively, try to compile in 32-bit mode on a regular 64-bit Linux. Then create a compressed file which is > 4 GB long (after compression). Now, try to `--list` it. See if `fseek()` survives or not. What happens when it doesn't.

2) In term of solution, there is `UTIL_fseek()` which is already available, and is _supposed_ to take care of such issue. I wouldn't mind a test to ensure that's effectively the case. If yes, the solution will be pretty simple : replace `fseek()` by `UTIL_fseek()`. Don't skip the test though : it is needed to show that there is a problem to fix, and that it is correctly fixed.

Reply:
Dropped `fseek` in favour of `UTIL_fseek`